File: simulation/simulate_random.py
import numpy as np
import matplotlib.pyplot as plt

# --- Binomial (manual) ---
def my_binom_1(n=1, p=1/3):
    u = np.random.rand(n)
    return np.sum(u < p)

print(my_binom_1(1000))
print(my_binom_1(1000, 0.5))

# --- Quantile Transform for Binomial ---
from scipy.stats import binom
def my_binom_2(n=1, p=1/3):
    u = np.random.rand()
    return binom.ppf(u, n, p)

print(my_binom_2(1000))
print(my_binom_2(1000, 0.5))

# --- Exponential via inverse CDF ---
u = np.random.rand(10000)
beta = 3
x = -beta * np.log(1 - u)
plt.hist(x, bins=30, color='lightblue')
plt.title("Exponential(β=3)")
plt.show()
print(np.mean(x))

# --- Gamma (accept-reject) ---
def ar_gamma(n=100):
    x = []
    while len(x) < n:
        u = np.random.rand()
        if u < 0.5:
            y = -np.log(1 - np.random.rand())
        else:
            y = -np.log(1 - np.random.rand()) - np.log(1 - np.random.rand())
        u2 = np.random.rand()
        temp = 2 * np.sqrt(y) / (1 + y)
        if u2 < temp:
            x.append(y)
    return np.array(x)

xg = ar_gamma(10000)
plt.hist(xg, bins=30, color='lightgreen')
plt.title("Gamma(3/2,1)")
plt.show()

print(np.mean(xg))

# --- Beta via rejection sampling ---
x1 = np.random.rand(300)
y1 = np.random.uniform(0, 2.6, 300)
from scipy.stats import beta
selected = y1 < beta.pdf(x1, 3, 6)
accepted = x1[selected]
plt.hist(accepted, bins=30, color='orange')
plt.title("Beta(3,6) via Rejection Sampling")
plt.show()

print(np.mean(selected))
print(np.mean(accepted < 0.5))
print(beta.cdf(0.5, 3, 6))

# --- Box–Muller Normal generation ---
u = np.random.rand(10000)
v = np.random.rand(10000)
r = np.sqrt(-2 * np.log(u))
theta = 2 * np.pi * v
x_norm = r * np.cos(theta)
plt.hist(x_norm, bins=40, color='skyblue')
plt.title("Box–Muller Normal(0,1)")
plt.show()

File: simulation/simulate_random.jl
using Random, Distributions, Plots

# --- Binomial (manual) ---
function my_binom_1(n::Int=1, p::Float64=1/3)
    u = rand(n)
    return sum(u .< p)
end
println(my_binom_1(1000))
println(my_binom_1(1000, 0.5))

# --- Quantile Transform ---
function my_binom_2(n::Int=1, p::Float64=1/3)
    u = rand()
    return quantile(Binomial(n, p), u)
end
println(my_binom_2(1000))
println(my_binom_2(1000, 0.5))

# --- Exponential via inverse CDF ---
u = rand(10000)
β = 3.0
x = -β .* log.(1 .- u)
histogram(x, bins=30, color=:lightblue, title="Exponential(β=3)")

println(mean(x))

# --- Gamma (accept-reject) ---
function ar_gamma(n::Int=100)
    x = Float64[]
    while length(x) < n
        u = rand()
        y = u < 0.5 ? -log(1 - rand()) : -log(1 - rand()) - log(1 - rand())
        u2 = rand()
        temp = 2 * sqrt(y) / (1 + y)
        if u2 < temp
            push!(x, y)
        end
    end
    return x
end

xg = ar_gamma(10000)
histogram(xg, bins=30, color=:lightgreen, title="Gamma(3/2,1)")
println(mean(xg))

# --- Beta via rejection sampling ---
x1 = rand(300)
y1 = rand(300) .* 2.6
selected = y1 .< pdf.(Beta(3,6), x1)
accepted = x1[selected]
histogram(accepted, bins=30, color=:orange, title="Beta(3,6) via Rejection Sampling")

println(mean(selected))
println(mean(accepted .< 0.5))
println(cdf(Beta(3,6), 0.5))

# --- Box–Muller Transform ---
u = rand(10000)
v = rand(10000)
r = sqrt.(-2 .* log.(u))
θ = 2π .* v
x_norm = r .* cos.(θ)
histogram(x_norm, bins=40, color=:lightblue, title="Box–Muller Normal(0,1)")
